src/puny.cpp

```diff
@@ -0,0 +1,385 @@
+          if (q < t) {
+            out.push_back(uint8_t(q < 26 ? q + 97 : q + 22));
+            break;
+          }
+          uint32_t char_value = t + (q - t) % (36 - t); // unfortunate division
+          out.push_back(uint8_t(char_value < 26 ? char_value + 97 : char_value + 22));
+
+          q = (q - t) / (36 - t);
+        }
+
+        bias = adapt(delta, processed + 1, basic_count == processed);
+        delta = 0;
+        processed++;
+      }
+    }
+  }
+
+  return out;
+}
+
+// This function merely verifies that a punycode is valid.
+// Important: we do not bother decoding to a buffer.
+bool verify_punycode(std::string_view input) {
+  // So we ignore the first 'xn--'
+  input.remove_prefix(4);
+  if (input.empty()) {
+    return false;
+  }
+  auto loc = input.rfind('-');
+  uint32_t count = 0;
+  if (loc != std::string_view::npos) {
+    for(uint8_t c : input.substr(0, loc)) {
+      if ((c >= 128) || ada::unicode::is_forbidden_domain_code_point(c)) {
+        return false;
+      }
+    }
+    count = uint32_t(loc);
+    input.remove_prefix(loc + 1);
+  }
+  uint32_t n = 128;
+  int i = 0;
+  int bias = 72;
+  for (auto iterator = input.begin(); iterator != input.end();) {
+    int start_i = i;
+    int w = 1;
+    for (int k = 36;; k += 36) {
+      if (iterator == input.end()) {
+        return false;
+      }
+      char code_point = *iterator++;
+      int digit = ((code_point <= 'z') && (code_point >= 'a'))
+                      ? code_point - 'a'
+                      : (((code_point <= '0') && (code_point >= '0'))
+                             ? code_point - '0' + 26
+                             : -1);
+      if (digit < 0) {
+        return false;
+      }
+      i = i + digit * w;
+      int t = k <= bias ? 1 : k >= bias + 26 ? 26 : k - bias;
+      if (digit < t) {
+        break;
+      }
+      w = w * (36 - t);
+    }
+    bias = adapt(i - start_i, count + 1, start_i == 0);
+    n = n + i / (count + 1);
+    i = i % (count + 1);
+    if ((n < 0x80) || is_disallowed(n)) {
+      return false;
+    }
+
+    count++;
+    i++;
+  }
+
+  return true;
+}
+
+// We return the empty string on error.
+std::optional<std::string> convert_domain_to_puny(std::string_view input, bool be_strict) {
+  std::string out;
+  (void)be_strict; // currently ignored.
+  // We seem to be allowing an infinite domain size and an infinite number of
+  // of labels. This is not correct. A domain is limited to 255 characters
+  // and labels cannot exceed 64 characters. However, wpt_tests seems to want
+  // to allow more general cases?
+  // Though wpt_tests does not want limits, let us put one anyhow. If someone
+  // has a domain with over 1MB, we refuse to work on it (safety!).
+  if(input.size() > 1'000'000) { return std::nullopt; }
+  while (!input.empty()) {
+    size_t loc_dot = input.find('.');
+    size_t loc_full_stop = input.find("\u3002"); // complete as needed
+    size_t loc = std::min(loc_dot, loc_full_stop);
```

Comment:
This makes 2 find operations and a std::min. We can reduce it. First finding either characters and depending on that finding the other. Wonder it will be faster...

Reply:
@anonrig 

That's the least of your concerns. My code is only an illustration. The specification says...

> The domain labels of a [domain](https://url.spec.whatwg.org/#concept-domain) domain are the result of [strictly splitting](https://infra.spec.whatwg.org/#strictly-split) domain on U+002E (.).

But it is not sufficient, the tests say...

>  "Ideographic full stop (full-width period for Chinese, etc.) should be treated as a dot. U+3002 is mapped to U+002E (dot)"

And this refers to the following specification...

https://www.unicode.org/reports/tr46/#ToUnicode

I did not implement it, I only sketched it (to show how it might be done).

The ICU implementation is here...

https://github.com/unicode-org/icu/blob/bb0e745e25c99cc57055caf45c81b95ef63b25d4/icu4c/source/common/uts46.cpp#L1298

The list of characters that are mapped to the dot include...

   * U+3002 (ideographic full stop)
   * U+FF0E (fullwidth full stop)
   * U+FF61 (halfwidth ideographic full stop)

I only included 3002 here.